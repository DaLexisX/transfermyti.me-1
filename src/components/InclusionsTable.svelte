<script lang="ts">
  import { createEventDispatcher } from "svelte";

  import type { BaseEntityModel, TableViewModel } from "~/typeDefs";
  import { booleanToYesNo } from "~/utilities/textTransforms";

  import InclusionsTableTitle from "~/components/InclusionsTableTitle.svelte";

  export let id: string;
  export let title: string;
  export let fieldNames: string[];
  export let fieldLabels: string[];
  export let tableRecords: TableViewModel<BaseEntityModel>[];
  export let totalCountsByType: Dictionary<number>;

  const dispatchEvent = createEventDispatcher();

  let canToggleAll: boolean;
  $: canToggleAll = tableRecords.some((record) => !record.existsInTarget);

  let totalCounts: number[];
  $: totalCounts = getTotalCounts(totalCountsByType);

  let fieldCount: number;
  $: fieldCount = fieldNames.length;

  function getTotalCounts(countsByType: Dictionary<number>): number[] {
    const { existsInTarget, ...validCountsByType } = countsByType;
    return Object.values(validCountsByType);
  }

  function handleIncludeButtonClick(): void {
    dispatchEvent("toggle-all");
  }

  function handleIncludeToggle(id: string): void {
    dispatchEvent("toggle-one", id);
  }
</script>

<InclusionsTableTitle
  id="{id}-title"
  {canToggleAll}
  on:toggle-all={handleIncludeButtonClick}
>
  {title}
</InclusionsTableTitle>

<table {id} role="grid" aria-labelledby="{id}-title">
  <thead>
    <tr>
      {#each fieldLabels as label}
        <th scope="col">{label}</th>
      {/each}

      <th data-include scope="col">Include?</th>
    </tr>
  </thead>

  <tbody>
    {#each tableRecords as record (record.id)}
      <tr data-disabled={record.existsInTarget}>
        {#each fieldNames as fieldName}
          <td>{booleanToYesNo(record[fieldName])}</td>
        {/each}

        <td data-include>
          <input
            type="checkbox"
            checked={record.existsInTarget ? false : record.isIncluded}
            disabled={record.existsInTarget}
            on:input={() => handleIncludeToggle(record.id)}
          />
        </td>
      </tr>
    {/each}
  </tbody>

  <tfoot>
    <tr>
      <th colspan={fieldCount + 1 - totalCounts.length}>
        {totalCounts.length === 1 ? "Total" : "Totals"}
      </th>

      {#each totalCounts as totalCount}
        <td>{totalCount}</td>
      {/each}
    </tr>
  </tfoot>
</table>
