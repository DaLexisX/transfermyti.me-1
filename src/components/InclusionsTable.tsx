import React from "react";

const InclusionsTable: React.FC = (props) => <table role="grid" {...props} />;

export default InclusionsTable;
