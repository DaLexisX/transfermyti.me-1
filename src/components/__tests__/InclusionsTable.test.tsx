import React from "react";

import InclusionsTable from "../InclusionsTable";
import { render } from "~/jestHelpers";

describe("the <InclusionsTable> component", () => {
  test("matches its snapshot with valid props", () => {
    const wrapper = render(<InclusionsTable />);

    expect(wrapper.container.firstElementChild).toMatchSnapshot();
  });
});
