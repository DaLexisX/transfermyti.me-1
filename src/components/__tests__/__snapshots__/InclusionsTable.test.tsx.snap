// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <InclusionsTable> component matches its snapshot with valid props 1`] = `
<table
  role="grid"
/>
`;
